### ISSUE TITLE
Requested Feature

### ISSUE THREAD

**Issue by texxasrulez** [blind-coder/rcmcarddav#295]
Can you integrate using wildcards for domains? Like "%d" "%s" etc. Your plugin does work using wildcards but displays the URL as "https://www.%d/carddav_location" . This is a feature request and not an issue I love your plugin. Thank you for your work and sharing that work.

**Comment by mstilkerich** [comment]
Hello,

  I'm not quite sure whether I understand the request correctly: Is this only about the URL being displayed in the preferences, not a functional change? %s is not a supported placeholder. You can use %u, %l and %d inside the URL to be substituted by the username or its local part / domain part in case the username is an e-mail address.

**Comment by texxasrulez** [comment]
Sometimes words just isn't enough so I will add an image as well after I explain a bit more.
I am using %d wildcard in the config file and I run into issues trying to sync from other devices. Trying to sync from other devices I get errors saying the URL cannot be found. They do not find my actual Domain name which %d should take care of but if I replace %d with my FQDN, all is right and works as expected. No errors are being thrown, so makes troubleshooting all that more hard. I may be doing something wrong but I figured I would ask before I get too deep into your plugin.

Screenshot of URL from Carddav Settings Page
![Wildcard-Domain](https://user-images.githubusercontent.com/10519182/86994155-9341c580-c15a-11ea-8a03-e9bc911e8307.png)

**Comment by mstilkerich** [comment]
Concerning the display: With v4, the URL of the addressbook will be stored (and therefore also displayed) without any placeholders, which I believe should address your original feature request.

Now as to your last comment, I take that the placeholder replacement is not working for you, also I'm again confused by your statement that it causes issues for "other devices". Other devices sync directly with the server, the roundcube plugin is not involved in this.

Maybe there is also a misunderstanding concerning what %d is replaced by: If your __roundcube username__ has the form of an email address (for example: gene@example.com), %d in the URL would be replaced by the domain part of that username (e. g. example.com). It is __not__ replaced with the FQDN of the server (which is a non-user-dependent constant that the admin can just directly insert).

**Comment by texxasrulez** [comment]
Sorry, I am a developer too and it is frustrating when users can't articulate what they want and expect and it seems I am one of those now, but let me try again. I do appreciate your patience.
What I want to do:
I am working on a plugin for i-MSCP to utilize several plugins with the main system roundcube instance. So with multiple domains required, and in respect to your carddav plugin, I was trying to use %d wildcard to convert to each users domain to their actual domain name but when using %d in config.inc.php for the FQDN portion of config, instead of a domain getting saved, it saves as https://www.%d.com/blahblah in the database and therefor cannot access actual contacts from Roundcube. The %u and %p wildcards work perfectly as expected but trying to get the wildcard %d for domain to work is not working as expected.

Actual config.inc.php url entry I am trying to use - "https://%d/cloud/remote.php/dav/addressbooks/users/%u/"

What I think should happen:
%d - domain (Not working as expected)
%u - username (works as expected)

Does this clarify for you? I am sorry if not. I do not want to take up too much of your time but I do love your plugin and want to expose it to more project and users.

**Comment by mstilkerich** [comment]
No worries, if there's a bug I'd like to fix it.

So what you describe all sounds reasonable except for the "it does not work" part. In RCMCardDAV 3.x, it is possible (but not guaranteed) that placeholders in the URL end up in the database. It depends on how the discovery processes. However, these placeholders are replaced whenever a request is sent to the server, so it should work just fine.

If you want, I'd suggest you try the current development version v4, installation instructions here:
https://github.com/blind-coder/rcmcarddav/tree/v4

I just tested a preset with the %d placeholder in the URL and it works fine. With v4, the URL is stored without placeholders in the database. If it still does not work for you, please provide the log files that are written to the logs/ subdirectory of roundcube, with names carddav.log and carddav_http.log. You may need to increase the log level to DEBUG in config.inc.php for those logs to contain enough information.

**Comment by texxasrulez** [comment]
PERFECT!  This works flawlessly and does exactly what I was wanting to achieve. No errors in my logs to report either I might add. Great work and it is greatly appreciated .

If you ever need any help don't hesitate to ask me. I can test the heck out of stuff ... I have no life so I got the time.
Thanks again and have a great day.

**Comment by mstilkerich** [comment]
Great to here it works, and thanks for the motivating words :-)
